### ISSUE TITLE
Accessibilty for internal UI

### ISSUE THREAD

**Issue by rb6502** [mamedev/mame#5357]
In reference to https://www.reddit.com/r/MAME/comments/ceitb9/blind_user_with_a_couple_questions/

The easiest route would probably be to add hooks so OSD can speak the name of the currently selected menu item using the Mac and Windows built-in speech synthesizers (and something like Festival or eSpeech on Linux).  This probably will best be done after Nathan finishes his UI overhaul though.

**Comment by MooglyGuy** [comment]
I disagree, since Nathan's UI overhaul effectively eliminates the internal UI in favor of a wxWidgets framework.

I was told not so long ago that any GPU hardware acceleration *must* have a software fallback for systems that lack 3D acceleration, because we apparently want to keep rendersw.hxx around as a fallback path. Just as much, it's important that MAME retain some sort of internally-rendered UI for systems that lack a viable wxWidgets port.

If I can sort out the Laserdisc problems that Tafoid reported, this sounds like something I'd actually rather like to look into this weekend, since I'm taking a vacation day on Friday.

**Comment by npwoods** [comment]
What MG said.  I'm bypassing the MAME UI, and I don't see how any
text-to-speech enhancements would collide with my changes.

While I am doing quite a bit of refactoring, my work involves trying to
better encapsulate the UI and addressing the knotty entanglements with the
core.

-N

**Comment by rb6502** [comment]
Ok.  I wasn't sure of the scope in terms of separating the existing UI's presentation from the business logic, but if you guys think it's OK and MG can do something this weekend I'm all for it.

**Comment by pitermach** [comment]
Hello,

I'm a blind Screen Reader user and I came here after wanting to open an issue about this exact topic and because I was just having a conversation about this with a few other people. There are many cases where having access to Mame's on-screen UI would be extremely helpful.

I initially came here to suggest [Accesskit](https://github.com/AccessKit/accesskit) to do this, which is a library that lets you give accessibility to UI controls on Windows, Mac and Linux even if they're just directly drawn to the screen like in the case of Mame. There is an example for making an SDL-based app accessible in the directory for Accesskit's C binding which may be helpful.

However, I'm now also wondering if the WX Widgets based UI is still something that's being considered? This would also solve the problem, because WX actually offers decent accessibility on its own so if the menus were changed to use it chances are the screen readers would be able to read them with little to no changes required.

**Comment by jaybird110127** [comment]
Accessibility for the MAME UI is something I, too, would really like to see. A few days ago, the author of an Apple IIe/c/gs Print/Braille word processing package for the blind passed away. I'd love to create a package to show people what using this program was like, but one of the major stumbling blocks is that once you get to a certain point, you have to turn the program disk over, which in MAME's case, means switching to another disk image. Currently this can't be done in the UI. You can sort of fake it using save states, but that's less than optimal, and you need to know what you're doing. Has there been any progress on this?

**Comment by rb6502** [comment]
I'm sorry nothing has happened with this yet.  MAME always has bigger ambitions than we end up having time to implement.  The suggestion of AccessKit looks interesting, I'm going to have to digest that a bit.

@jaybird110127 Perhaps the program could be adapted to run from a larger volume like an 3.5 inch disk or hard drive to avoid the disk swapping?  If you can point us to the program I suspect @peterferrie could take a look.

**Comment by sevapopov2** [comment]
Hello everyone,
Is there now any way to access Mame's settings to assign keyboard or controller keys for playing?
Currently when pressing Tab in the game the interface is not accessible.
I need to assign keys to be able to play UMK3 which uses Mame emulator and Kaillera for netplay.
Thank you!

**Comment by jaybird110127** [comment]
This issue has come up again for me. A few days ago I learned that MAME now emulates the Yamaha PortaSound PSS-170 keyboard. Imagine my disappointment when I booted it up, and could do absolutely nothing whatsoever with it because apparently there are no keyboard assignments at all. On Friday night I was able to use OCR in a rather crude way to assign some random control to the right control key, then manually edit the XML configuration and change that assignment to the proper control for Demo Start/Stop, by referencing the source code. Now I can start or stop the demo, but that's it. This has been a sticking point for years, as there's no good way for me to even learn what the keyboard assignments are on a new system.

**Comment by hwangsihu** [comment]
Hi,
I first encountered classic games on a console, but since it is now broken, I am using MAME.
I enjoy playing fighting games, but I am not skilled enough to play characters other than the bosses, so I have to use cheats to activate them. It is quite a hassle to press the Tab key, then the arrow keys more than 10 times, press Enter, and ask the AI ​​for the menu order every time.
I would be very happy if a visually impaired person could access the menus without any additional tools.
Thank you.
